Lib/test/test_unittest/testmock/testhelpers.py

```diff
@@ -1107,6 +1107,32 @@ class WithNonFields:
                 with self.assertRaisesRegex(AttributeError, msg):
                     mock.b
 
+    def test_dataclass_default_value_type_overrides_field_annotation(self):
+        # If field defines an actual default, we don't need to change
+        # the default type. Since this is how it used to work before #124176
+        @dataclass
+        class WithUnionAnnotation:
+            narrow_default: int | None = field(default=30)
+
+        for mock in [
+            create_autospec(WithUnionAnnotation, instance=True),
+            create_autospec(WithUnionAnnotation()),
+        ]:
+            with self.subTest(mock=mock):
+                self.assertIs(mock.narrow_default.__class__, int)
```

Comment:
I feel both of these tests would be clearer if they actually showed what the thing ends up being, not just it's type, something like:

```suggestion
                self.assertIs(mock.narrow_default.__class__, int)
                self.assertEqual(mock.narrow_default, 30)
```

Reply:
Adding something along these lines would also distinguish the cases where we're expecting to mock a specific value from those where we're deriving an instance spec from a declared runtime type